mora/service/engagement.py

```diff
@@ -231,6 +229,70 @@ def get_title_id(effect):
         except (KeyError, IndexError):
             return None
 
+    def convert_engagement(funcid, effect, start, end):
+        return {
+            "uuid": funcid,
+
+            keys.PERSON: user_cache[get_employee_id(effect)],
+            keys.ORG_UNIT: unit_cache[get_unit_id(effect)],
+            keys.JOB_FUNCTION: class_cache[get_title_id(effect)],
+            keys.ENGAGEMENT_TYPE: class_cache[get_type_id(effect)],
+
+            keys.VALIDITY: {
+                keys.FROM: util.to_iso_time(start),
+                keys.TO: util.to_iso_time(end),
+            }
+        }
+
+    def convert_association(funcid, effect, start, end):
+        return {
+            "uuid": funcid,
+
+            keys.PERSON: user_cache[get_employee_id(effect)],
+            keys.ORG_UNIT: unit_cache[get_unit_id(effect)],
+            keys.JOB_FUNCTION: class_cache[get_title_id(effect)],
+            keys.ASSOCIATION_TYPE: class_cache[get_type_id(effect)],
+
+            keys.VALIDITY: {
+                keys.FROM: util.to_iso_time(start),
+                keys.TO: util.to_iso_time(end),
+            }
+        }
+
+    def convert_role(funcid, effect, start, end):
+        return {
+            "uuid": funcid,
+
+            keys.PERSON: user_cache[get_employee_id(effect)],
+            keys.ORG_UNIT: unit_cache[get_unit_id(effect)],
+            keys.ROLE_TYPE: class_cache[get_type_id(effect)],
+
+            keys.VALIDITY: {
+                keys.FROM: util.to_iso_time(start),
+                keys.TO: util.to_iso_time(end),
+            }
+        }
+
+    def convert_leave(funcid, effect, start, end):
+        return {
+            "uuid": funcid,
+
+            keys.PERSON: user_cache[get_employee_id(effect)],
+            keys.LEAVE_TYPE: class_cache[get_type_id(effect)],
+
+            keys.VALIDITY: {
+                keys.FROM: util.to_iso_time(start),
+                keys.TO: util.to_iso_time(end),
+            }
+        }
+
+    converters = {
+        'engagement': convert_engagement,
+        'association': convert_association,
+        'role': convert_role,
+        'leave': convert_leave,
+    }
+
```

Comment:
Looks fine for now, but I believe we can make it even more general — we'll eventually want to have this fully configurable, I believe.

The validity assignment seems redundant, though; perhaps you could move it to the caller?

Reply:
I agree - I thought about how to make this fully configurable and eventually expand on my rudimentary mappings and include them in this.
It's a somewhat non-trivial task, so I settled on this for now, just to actually have it working.

I agree on your validity point as well. I'll get it done.